### PR TITLE
Gun adapter

### PR DESCRIPTION
**Pull request by alex-strizhakov**
Expanding gun adapter:
- unnecessary sni removed
- added certificates verification
- added support for socks proxy
- bug fixes for veryfy_fun and closing conn
- updated docs
- errors namings
- some code refactor

### CODE REVIEW

File: lib/tesla/adapter/gun.ex [hunk 1]
```diff
@@ -216,6 +237,36 @@ if Code.ensure_loaded?(:gun) do
...
+      received_version =
+        to_string(proxy_type)
+        |> String.last()
+
+      version =
+        if received_version in ["4", "5"] do
+          String.to_integer(received_version)
+        else
+          5
+        end
```
Comment (chvanikoff) [inline lib/tesla/adapter/gun.ex]:
```elixir
version =
  proxy_type
  |> to_string()
  |> String.last()
  |> case do
    "4" -> 4
    _ -> 5
  end
```

File: test/tesla/adapter/gun_test.exs [hunk 2]
```diff
@@ -237,17 +305,37 @@ defmodule Tesla.Adapter.GunTest do
...
+      {:fin, body} ->
         acc <> body
 
       {:nofin, part} ->
-        read_body(pid, stream, acc <> part, close_conn)
+        read_body(pid, stream, opts, acc <> part)
     end
   end
+
+  defp reduce_body(body), do: Enum.reduce(body, "", fn part, acc -> acc <> part end)
```
Comment (chvanikoff) [inline test/tesla/adapter/gun_test.exs]: `Enum.join/1`?

Comment (alex-strizhakov) [reply test/tesla/adapter/gun_test.exs]: 😅

File: lib/tesla/adapter/shared.ex [hunk 3]
```diff
@@ -12,4 +12,13 @@ defmodule Tesla.Adapter.Shared do
...
   defp parse_chunk(_), do: :eof
+
+  @spec prepare_path(String.t() | nil, String.t() | nil) :: String.t()
+  def prepare_path(nil, nil), do: "/"
+  def prepare_path(nil, query), do: "/?" <> query
+  def prepare_path(path, nil), do: path
+  def prepare_path(path, query), do: path <> "?" <> query
+
+  @spec format_method(atom()) :: String.t()
+  def format_method(method), do: to_string(method) |> String.upcase()
```
Comment (teamon) [inline lib/tesla/adapter/shared.ex]: Are these functions used in any other adapter? Uppercase binary method seems gun-specific for example.

File: lib/tesla/adapter/gun.ex [hunk 4]
```diff
@@ -356,13 +397,16 @@ if Code.ensure_loaded?(:gun) do
...
+            read_body(pid, stream, opts, acc)
           end
+
+        {:DOWN, _, _, _, reason} ->
+          if opts[:close_conn], do: close(pid)
+          {:error, reason}
       after
         opts[:timeout] || @adapter_default_timeout ->
-          {:error, :timeout}
+          {:error, :recv_body_timeout}
```
Comment (teamon) [inline lib/tesla/adapter/gun.ex]: No close here?

File: lib/tesla/adapter/shared.ex (hunk 3 above)
Comment (alex-strizhakov) [reply lib/tesla/adapter/shared.ex]: Mint adapter will be upstreamed soon. These functions are used there too.

File: lib/tesla/adapter/gun.ex (hunk 4 above)
Comment (alex-strizhakov) [reply lib/tesla/adapter/gun.ex]: yes, this `:error` case will be closed in `format_response/6` with `:plain` clause. Also i removed unnecessary close check in `{:DOWN, _, _, _, reason}` clause.

File: lib/tesla/adapter/shared.ex (hunk 3 above)
Comment (teamon) [reply lib/tesla/adapter/shared.ex]: 👌

**Pull request merged**